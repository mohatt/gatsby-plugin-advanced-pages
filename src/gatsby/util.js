import fs from 'fs'
import path from 'path'
import _ from 'lodash'
import debug from 'debug'

// Initializes options
// The function only runs when called with args to initialize and cache
// options object. Subsequent calls without args return the cached value
export function initializeOptions (args) {
  if (arguments.length === 0) {
    if (typeof initializeOptions.options === 'undefined') {
      return reportError('Cant fetch options because they are not initialized')
    }

    return initializeOptions.options
  }

  // Merge user-defined options with defaults
  const options = _.merge(args.defaultOptions, args.pluginOptions)

  // set the root path of the the project
  options._root = args.store.getState().program.directory

  // Ensure basePath is absolute
  options.basePath = path.join('/', options.basePath)

  // Cache the options object now so that subsequent
  // calls dont throw an error
  initializeOptions.options = options

  // Verify directory locations
  options.directories = _.mapValues(options.directories, dir => {
    return lookupPath(dir)
  })

  // Verify default template
  if(options.template) {
    options.template = lookupPath(options.template, options.directories.templates)
  }

  // Debug final options object
  debug('gatsby-plugin-advanced-pages')('Options', options)

  initializeOptions.options = options
}

// Initializes reporter
// The function only runs when called with args to initialize and cache
// reporter object. Subsequent calls without args return the cached value
export function initializeReporter (reporter) {
  if (arguments.length === 0) {
    if (typeof initializeReporter.reporter === 'undefined') {
      throw new Error('Cant fetch reporter because it is not initialized')
    }

    return initializeReporter.reporter
  }

  const errorMap = {
    [10000]: {
      text: context => context.message,
      level: `ERROR`,
      type: `PLUGIN`,
    },
  }

  if (reporter.setErrorMap) {
    reporter.setErrorMap(errorMap)
  }

  initializeReporter.reporter = reporter
}

// Gets the initialized eporter object
export function getReporter () {
  return initializeReporter()
}

// Gets the initialized eporter object
export function reportError (message, e = null) {
  const reporter = getReporter()
  const prefix = `"gatsby-plugin-advanced-pages" threw an error while running`

  if(!e) {
    reporter.panic({
      id: `10000`,
      context: {
        message: prefix
      },
      error: new Error(message)
    })
    return
  }

  reporter.panic({
    id: `10000`,
    context: {
      message: `${prefix}:\n ${message}`
    },
    error: e
  })
}

// Gets the initialized options object
export function getOptions () {
  return initializeOptions()
}

// Gets the value of a single option
export function getOption (optionName) {
  return _.get(getOptions(), optionName)
}

// Checks if a location exists under parent
// if not checks if it exists under project root
// if not check if its an absolute path
// throws an error if cant find any
export function lookupPath(location, parent = null){
  if(!location || typeof location !== 'string') {
    return reportError(
      `ensurePath() expected a non-empty string but got ${typeof location}("${location}")`
    )
  }

  let search = []
  if(parent) {
    const localPath = path.join(parent, location)
    search.push(localPath)
    if(fs.existsSync(localPath)) {
      return localPath
    }
  }

  const rootPath = path.join(getOption('_root'), location)
  search.push(rootPath)
  if(fs.existsSync(rootPath)) {
    return rootPath
  }

  search.push(location)
  if(fs.existsSync(location)) {
    return location
  }

  reportError(
    `A path with value "${location}" could not be found at ` +
    `any of the following locations:\n - "${search.join(`\n - "`)}"`
  )
}
