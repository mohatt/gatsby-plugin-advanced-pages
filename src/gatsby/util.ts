import fs from 'fs'
import path from 'path'
import _ from 'lodash'
import debug from 'debug'
import type { NodePluginArgs, Reporter } from 'gatsby'
import type { PluginOptions } from './types'

export interface ResolvedPluginOptions extends PluginOptions {
  _root: string;
}

/**
 * Represents the internal state of the plugin.
 */
const internalState: {
  options?: ResolvedPluginOptions
  reporter?: Reporter
} = {}

// Initializes options
// The function only runs when called with args to initialize and cache
// options object. Subsequent calls without args return the cached value
export const initializeOptions = (args?: Pick<NodePluginArgs, 'store'> & { pluginOptions: Partial<PluginOptions>, defaultOptions: PluginOptions }): ResolvedPluginOptions => {
  if (args === undefined) {
    if (!('options' in internalState)) {
      reportError('Cant fetch options because they are not initialized')
      return null
    }

    return internalState.options
  }

  // Merge user-defined options with defaults
  const options: ResolvedPluginOptions = _.merge(
    // set the root path of the project
    { _root: args.store.getState().program.directory },
    args.defaultOptions,
    args.pluginOptions,
  )
  // Ensure basePath is absolute
  options.basePath = path.join('/', options.basePath)
  // Ensure type names are Title case
  options.typeNames = _.mapValues(options.typeNames, _.upperFirst)
  // Convert directory paths to absolute ones
  options.directories = _.mapValues(options.directories, dir => (
    path.join(options._root, dir)
  ))

  // Cache the options object now so that subsequent
  // calls to lookupPath() dont throw an error
  internalState.options = options

  // Verify default template
  if (options.template) {
    options.template = lookupPath(options.template, options.directories.templates)
  }

  // Debug final options object
  debug('gatsby-plugin-advanced-pages')('Options', options)

  internalState.options = options

  return options
}

// Initializes reporter
// The function only runs when called with args to initialize and cache
// reporter object. Subsequent calls without args return the cached value
export const initializeReporter = (reporter?: Reporter) => {
  if (reporter === undefined) {
    if (!('reporter' in internalState)) {
      throw new Error('Cant fetch reporter because it is not initialized')
    }

    return internalState.reporter
  }

  if (reporter.setErrorMap) {
    reporter.setErrorMap({
      10000: {
        text: context => context.message,
        level: 'ERROR',
        type: 'PLUGIN'
      }
    })
  }

  internalState.reporter = reporter
  return reporter
}

// Gets the initialized reporter object
export const getReporter = () => {
  return initializeReporter()
}

// Prints an error message and terminates the build
export const reportError = (message: string, error?: Error): never => {
  const reporter = getReporter()
  const prefix = '"gatsby-plugin-advanced-pages" threw an error while running'

  if (!error) {
    return reporter.panic({
      id: '10000',
      context: {
        message: prefix
      },
      error: new Error(message)
    })
  }

  return reporter.panic({
    id: '10000',
    context: {
      message: `${prefix}:\n ${message}`
    },
    error
  })
}

// Prints a warning message
export const reportWarning = (message: string) => {
  const reporter = getReporter()
  const prefix = '"gatsby-plugin-advanced-pages" might not be working properly'
  reporter.warn(`${prefix}:\n ${message}`)
}

// Gets the initialized options object
export const getOptions = () => {
  return initializeOptions()
}

// Gets the value of a single option
export const getOption = <T extends keyof ResolvedPluginOptions>(optionName: T): ResolvedPluginOptions[T] => {
  return _.get(getOptions(), optionName)
}

// Checks if a location exists under parent
// if not checks if it exists under project root
// if not check if its an absolute path
// throws an error if cant find any
export const lookupPath = (location: string, parent?: string) => {
  if (!location || typeof location !== 'string') {
    reportError(
      `ensurePath() expected a non-empty string but got ${typeof location}("${location}")`
    )
    return null
  }

  const search = []
  if (parent && typeof parent === 'string') {
    const localPath = path.join(parent, location)
    search.push(localPath)
    if (fs.existsSync(localPath)) {
      return localPath
    }
  }

  const rootPath = path.join(getOption('_root'), location)
  search.push(rootPath)
  if (fs.existsSync(rootPath)) {
    return rootPath
  }

  if (path.isAbsolute(location)) {
    search.push(location)
    if (fs.existsSync(location)) {
      return location
    }
  }

  reportError(
    `A path with value "${location}" could not be found at ` +
    `any of the following locations:\n - "${search.join('\n - "')}"`
  )
  return null
}
