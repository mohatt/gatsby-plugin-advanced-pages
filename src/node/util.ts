import fs from 'fs'
import path from 'path'
import { createRequire } from 'module'
import Yaml from 'js-yaml'
import { validateOptionsSchema, Joi, PluginOptionsSchemaJoi, Schema } from 'gatsby-plugin-utils'
import _ from 'lodash'
import debug from 'debug'
import type { GatsbyNode, NodePluginArgs, Reporter } from 'gatsby'
import type { DefaultPluginOptions, PluginOptions } from './api'

/**
 * Represents the resolved internal options for the plugin.
 * It includes additional properties required for internal processing.
 */
interface ResolvedPluginOptions extends DefaultPluginOptions {
  /**
   * Absolute path to the root of the Gatsby project.
   * Used to resolve relative paths for templates, directories, etc.
   */
  _root: string
}

/**
 * Provides functionality to initialize, merge, and retrieve plugin options.
 *
 * @internal
 */
export const options = (() => {
  let ref: ResolvedPluginOptions | undefined

  /**
   * Initializes and merges plugin options with defaults.
   */
  const initialize = (args: {
    store: NodePluginArgs['store']
    pluginOptions: PluginOptions
    defaultOptions: DefaultPluginOptions
  }): ResolvedPluginOptions => {
    // Merge user-defined options with defaults
    const opts: ResolvedPluginOptions = _.merge(
      // set the root path of the project
      { _root: args.store.getState().program.directory },
      args.defaultOptions,
      args.pluginOptions,
    )
    // Normalize options
    Object.assign(opts, {
      // Ensure basePath is absolute
      basePath: path.join('/', opts.basePath),
      // Ensure type names are Title case
      typeNames: _.mapValues(opts.typeNames, _.upperFirst),
      // Convert directory paths to absolute ones
      directories: _.mapValues(opts.directories, (dir) => path.join(opts._root, dir)),
    })

    // Cache the options object now so that subsequent
    // calls to ensurePath() don't throw an error
    ref = opts

    // Verify default template exists
    if (opts.template != null) {
      opts.template = ensurePath(opts.template, opts.directories.templates)
    }

    // Debug final options object
    debug('gatsby-plugin-advanced-pages')('Options', opts)

    ref = opts

    return opts
  }

  /**
   * Retrieves a specific option value or the entire options object.
   *
   * - If called without arguments, returns the entire options object.
   * - If called with a key, retrieves the corresponding value.
   * - Throws an error if options are not initialized.
   *
   * @param optionName (Optional) The key of the option to retrieve.
   *
   * @returns The requested option value or the entire options object.
   *
   * @throws - Terminates execution if options are not initialized.
   */
  const get: {
    (): ResolvedPluginOptions
    <T extends keyof ResolvedPluginOptions>(optionName: T): ResolvedPluginOptions[T]
  } = (...args: [] | [string]) => {
    if (ref === undefined) {
      // throw new Error('Plugin options are not initialized.')
      throw new PluginError('Cant fetch options because they are not initialized')
    }

    if (args.length === 0) {
      return ref
    }

    return _.get(ref, ...args)
  }

  return { initialize, get }
})()

/**
 * Manages Gatsby's reporter instance, providing functionalities for structured error logging,
 * warnings, and instance management.
 *
 * @internal
 */
export const reporter = (() => {
  let ref: Reporter | undefined

  // Initializes reporter
  /**
   * Initializes and caches the Gatsby reporter instance.
   *
   * If `setErrorMap` is available, it registers custom error messages
   * for better debugging in Gatsby's CLI.
   */
  const initialize = (instance: Reporter): Reporter => {
    if (instance.setErrorMap != null) {
      instance.setErrorMap({
        10000: {
          text: (context) => context.message,
          category: 'USER',
          level: 'ERROR',
          type: 'PLUGIN',
        },
      })
    }

    ref = instance
    return instance
  }

  /**
   * Retrieves the cached reporter instance.

   * @throws Error if the reporter is accessed before being initialized.
   */
  const get = (): Reporter => {
    if (ref === undefined) {
      throw new Error('Cant fetch reporter instance because it is not initialized')
    }

    return ref
  }

  /**
   * Logs an error and terminates the build process.
   *
   * - Uses Gatsby's `panic()` method to log structured errors.
   * - If an `Error` object is provided, includes it in the error report.
   * - Otherwise, creates a new error instance from the given message.
   *
   * @param api The Gatsby Node API that caused the error.
   *
   * @param titleOrError The error message or PluginError object to display.
   *
   * @param error (Optional) The original error instance.
   *
   * @throws - This function never returns; it always terminates execution.
   */
  const error = (
    api: keyof GatsbyNode,
    titleOrError: string | PluginError | Error,
    error?: Error,
  ): never => {
    const instance = get()
    const prefix = `"gatsby-plugin-advanced-pages" threw an error during "${api}" hook`

    let title: string | undefined
    let mainError: Error | undefined

    if (titleOrError instanceof PluginError) {
      title = titleOrError.message
      mainError = titleOrError.originalError
    } else if (titleOrError instanceof Error) {
      mainError = titleOrError
    } else {
      title = titleOrError
    }

    if (error) {
      mainError = error
    }

    return instance.panic({
      id: '10000',
      context: {
        message: mainError && title ? `${prefix}:\n ${title}` : prefix,
      },
      error: mainError ?? (title ? new Error(title) : undefined),
    })
  }

  /**
   * Logs a warning message in the Gatsby CLI.
   *
   * @param message The warning message to display.
   */
  const warning = (message: string): void => {
    const instance = get()
    const prefix = '"gatsby-plugin-advanced-pages" might not be working properly'
    instance.warn(`${prefix}:\n ${message}`)
  }

  return { initialize, get, error, warning }
})()

/**
 * Validates the input data using Joi schema validation.
 *
 * @param schema The Joi schema (or schema factory) to validate against.
 * @param data The input data to validate.
 * @param [name] (Optional) The prop name of the data being validated. Default is `data`.
 *
 * @returns A promise that resolves to the validated and typed data.
 *
 * @throws {ValidationError} Throws an error if the input does not meet the schema requirements.
 */
export const validateSchema = async <E>(
  schema: Schema | ((Joi: PluginOptionsSchemaJoi) => Schema),
  data: unknown,
  name = 'data',
): Promise<E> => {
  const objSchema = Joi.object({ [name]: schema instanceof Function ? schema(Joi) : schema })
  const { value } = await validateOptionsSchema(objSchema, { [name]: data })
  return value[name] as E
}

/**
 * Ensures a path exists based on different search locations:
 *
 * The function checks in the following order:
 * 1. If `location` exists under the given `parent` directory, return it.
 * 2. If not found, checks under the Gatsby project root (`_root`).
 * 3. If still not found and `location` is absolute, it verifies the path exists.
 * 4. If none of the above match, logs an error and terminates execution.
 *
 * @param location The relative or absolute file path to resolve.
 * @param parent (Optional) A parent directory to check first.
 *
 * @returns The absolute path if found; otherwise, throws an error.
 *
 * @throws PluginError if the path is not found.
 */
export const ensurePath = (location: string, parent?: string): string => {
  const search = []
  if (parent != null) {
    const localPath = path.join(parent, location)
    search.push(localPath)
    if (fs.existsSync(localPath)) {
      return localPath
    }
  }

  const rootPath = path.join(options.get('_root'), location)
  search.push(rootPath)
  if (fs.existsSync(rootPath)) {
    return rootPath
  }

  if (path.isAbsolute(location)) {
    search.push(location)
    if (fs.existsSync(location)) {
      return location
    }
  }

  throw new PluginError(
    `A path with value "${location}" could not be found at ` +
      `any of the following locations:\n - "${search.join('\n - "')}"`,
  )
}

/**
 * Retrieves the default export of a module.
 */
export const getDefaultExport = async <T = unknown>(modulePath: string): Promise<T> => {
  const noop = {}
  let defaultExport: any = noop

  // Avoid require in tests since Vitest doesn't support mocking require calls
  if (process.env.VITEST !== 'true') {
    const nodeRequire = require ?? createRequire(import.meta.url)
    try {
      defaultExport = nodeRequire(modulePath)
      if (defaultExport?.__esModule === true && 'default' in defaultExport) {
        defaultExport = defaultExport.default
      }
    } catch (error) {
      if (error.code !== 'ERR_REQUIRE_ESM') {
        throw new PluginError(`Unable to import module "${modulePath}"`, error)
      }
    }
  }

  if (defaultExport === noop) {
    try {
      const esmExports = await import(modulePath)
      defaultExport = esmExports.default
    } catch (error) {
      throw new PluginError(`Unable to import ESM module "${modulePath}"`, error)
    }
  }

  if (defaultExport == null) {
    throw new PluginError(
      `Unable to import module "${modulePath}"`,
      new Error(`Couldn't find valid default export`),
    )
  }

  return defaultExport as T
}

/**
 * Searches for a configuration file under the given root directory and loads its content.
 *
 * @param root The root directory to search in.
 * @param baseName The base name of the config file (e.g., `"pages.config"`).
 *
 * @returns An object containing the file path and its parsed data, or `null` if not found.
 */
export const findConfigFile = async (
  root: string,
  baseName: string,
): Promise<{ path: string; data: unknown } | null> => {
  const extensions = ['yaml', 'yml', 'json', 'js', 'cjs', 'mjs']

  for (const ext of extensions) {
    const filePath = path.join(root, `${baseName}.${ext}`)
    if (!fs.existsSync(filePath)) {
      continue
    }

    try {
      switch (ext) {
        case 'yaml':
        case 'yml':
          return {
            path: filePath,
            data: Yaml.load(fs.readFileSync(filePath, 'utf8'), {
              filename: filePath,
              schema: Yaml.FAILSAFE_SCHEMA,
            }),
          }

        case 'json':
          return {
            path: filePath,
            data: JSON.parse(fs.readFileSync(filePath, 'utf8')),
          }

        default:
          return { path: filePath, data: await getDefaultExport(filePath) }
      }
    } catch (error) {
      throw new PluginError(`Failed to load config file "${filePath}"`, error)
    }
  }

  return null
}

/**
 * Wraps a Gatsby API function with error handling.
 *
 * @param api - The name of the Gatsby API being wrapped (e.g., `"onCreatePage"`).
 * @param fn - The original Gatsby API function to be wrapped.
 *
 * @returns The wrapped function with error handling.
 */
export const createPluginExport = <T extends keyof GatsbyNode, F extends GatsbyNode[T]>(
  api: T,
  fn: F,
): F => {
  return ((...args: any[]) => {
    try {
      // Call the original function with provided arguments
      const result = (fn as any)(...args)

      // If the function returns a Promise, attach error handling
      if (result instanceof Promise) {
        return result.catch((error) => reporter.error(api, error))
      }

      // Return the result (for synchronous functions)
      return result
    } catch (error) {
      // Catch synchronous errors and report them
      reporter.error(api, error)
    }
  }) as F
}

/**
 * Custom error class for plugin-related errors.
 */
export class PluginError extends Error {
  /**
   * @param message - A description of the error.
   * @param originalError - The original error that caused this issue (optional).
   */
  constructor(
    message: string,
    public readonly originalError?: Error,
  ) {
    super(message)
  }
}
