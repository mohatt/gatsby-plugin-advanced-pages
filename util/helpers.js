const path = require('path')
const debug = require("debug")("gatsby-plugin-advanced-pages")

// Checks if a path is a child of another one
function isPathChildof (child, parent) {
  const relative = path.relative(parent, child)
  return relative && !relative.startsWith('..') && !path.isAbsolute(relative)
}

module.exports = {
  debug,
  isPathChildof,
}
