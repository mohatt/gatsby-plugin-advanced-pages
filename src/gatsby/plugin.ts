import type { GatsbyNode, PluginOptions as GatsbyPluginOptions } from 'gatsby'
import { options, reporter } from './util'
import { getOptionsSchema } from './schema'
import type { PluginOptions } from './index'

// Validates user-defined options against schema
// runs before onPreBootstrap
export const pluginOptionsSchema: GatsbyNode['pluginOptionsSchema'] = ({ Joi }) => {
  return getOptionsSchema(Joi)
}

// Initializes plugin state
export const onPluginInit: GatsbyNode['onPluginInit'] = (
  args,
  pluginOptions: GatsbyPluginOptions & PluginOptions,
) => {
  // Default values for options
  const defaultOptions: PluginOptions = {
    basePath: '/',
    pages: [],
    template: null,
    directories: {
      templates: './src/templates',
      helpers: './gatsby/pages',
    },
    pagination: {
      limit: 10,
      suffix: '/page/:page',
    },
    typeNames: {
      page: 'Page',
      pageRoute: 'PageRoute',
    },
  }

  // Initializes and validates options
  // Only runs once at bootstrap
  options.initialize({
    store: args.store,
    defaultOptions,
    pluginOptions,
  })

  // Initializes plugin reporter
  reporter.initialize(args.reporter)
}

// Creates a webpack alias for the plugin cache directory
// Usually located at <root>/.cache/caches/<plugin>
export const onCreateWebpackConfig: GatsbyNode['onCreateWebpackConfig'] = ({ actions, cache }) => {
  actions.setWebpackConfig({
    resolve: {
      alias: {
        'gatsby-plugin-advanced-pages-cache': cache.directory,
      },
    },
  })
}

export const createSchemaCustomization: GatsbyNode['createSchemaCustomization'] = ({
  actions,
  schema,
}) => {
  const { createTypes } = actions
  const { page, pageRoute } = options.get('typeNames')

  createTypes([
    schema.buildObjectType({
      name: pageRoute,
      fields: {
        name: 'String!',
        path: 'String!',
      },
    }),
    schema.buildObjectType({
      name: page,
      fields: {
        title: 'String!',
        templateName: 'String',
        template: 'String',
        helper: 'String',
        data: 'JSON',
        routes: `[${pageRoute}!]!`,
      },
      interfaces: ['Node'],
    }),
  ])
}

export { default as sourceNodes } from './source-nodes'
export { default as createPages } from './create-pages'
