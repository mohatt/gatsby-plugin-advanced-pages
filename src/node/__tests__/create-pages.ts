import path from 'path'
import { setupPlugin, mountDir, mountModule, readFile, resetVFS } from '@test/util'
import testCases from './__fixtures__/create-pages'
import { createPages } from '../plugin'

describe('createPages', () => {
  const graphql = vi.fn()
  const createPage = vi.fn()
  const cache = { directory: '/path/to/.cache' }
  const actions = { createPage }
  const getNodesByType = vi.fn().mockReturnValue({ length: 10 })
  const args: any = {
    graphql,
    actions,
    cache,
    getNodesByType,
  }
  const helperFile = '/path/to/helper.js'

  beforeEach(() => {
    setupPlugin()
    resetVFS()
    graphql.mockReset()
    createPage.mockReset()
    getNodesByType.mockClear()
  })

  it('correctly calls graphql pages query', async () => {
    graphql.mockReturnValue(Promise.reject(new Error('Test error')))
    await expect(createPages(args)).rejects.toThrow()
    expect(graphql.mock.calls).toMatchSnapshot()
  })

  it.each(testCases)('$title', async ({ throws, pages, helper }) => {
    if (helper !== undefined) mountModule(helperFile, helper)

    // Create a virtual directory for cache.directory
    // so that routes.json file can be written virtually
    mountDir(cache.directory)

    graphql.mockReturnValue(
      Promise.resolve({
        data: {
          allPage: { nodes: pages },
        },
      }),
    )

    const promise = expect(createPages(args))
    if (throws) {
      await promise.rejects.toMatchSnapshot('error')
      return
    }

    await promise.resolves.toBeUndefined()
    expect(createPage.mock.calls).toMatchSnapshot('createPage')
    expect(readFile(path.join(cache.directory, 'routes.json'))).toMatchSnapshot('routes.json')
  })

  it('correctly handles graphql result errors', async () => {
    const errors = [new Error('Some error')]
    graphql.mockReturnValue(Promise.resolve({ errors }))
    await expect(createPages(args)).rejects.toMatchSnapshot()
  })

  it('correctly handles graphql errors', async () => {
    graphql.mockReturnValue(Promise.reject(new Error('Some fatal error')))
    await expect(createPages(args)).rejects.toMatchSnapshot()
  })

  it('correctly handles empty getNodesByType result', async () => {
    getNodesByType.mockReturnValueOnce({ length: 0 })
    await expect(createPages(args)).resolves.toBeUndefined()
  })

  it('correctly handles routeMap write errors', async () => {
    graphql.mockReturnValue(
      Promise.resolve({
        data: {
          allPage: { nodes: [] },
        },
      }),
    )
    await expect(createPages(args)).rejects.toMatchSnapshot()
  })
})
