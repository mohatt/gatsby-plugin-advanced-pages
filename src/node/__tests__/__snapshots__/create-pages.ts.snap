// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createPages > 'correctly creates no pages' > createPage 1`] = `[]`;

exports[`createPages > 'correctly creates no pages' > routes.json 1`] = `"[]"`;

exports[`createPages > 'correctly creates pages (blog-paginated-auto-route)' > createPage 1`] = `
[
  [
    {
      "component": "/path/to/blog.js",
      "context": {
        "id": "blog-id",
        "limit": 10,
        "offset": 0,
      },
      "path": "/blog",
    },
  ],
  [
    {
      "component": "/path/to/blog.js",
      "context": {
        "id": "blog-id",
        "limit": 10,
        "offset": 10,
      },
      "path": "/blog/page/2",
    },
  ],
  [
    {
      "component": "/path/to/blog.js",
      "context": {
        "id": "blog-id",
        "limit": 10,
        "offset": 20,
      },
      "path": "/blog/page/3",
    },
  ],
]
`;

exports[`createPages > 'correctly creates pages (blog-paginated-auto-route)' > routes.json 1`] = `"[{"name":"blog","path":"/blog","scopes":{"pagination":"/blog/page/:page"}}]"`;

exports[`createPages > 'correctly creates pages (blog-paginated-custom-limit)' > createPage 1`] = `
[
  [
    {
      "component": "/path/to/blog.js",
      "context": {
        "id": "blog-id",
        "limit": 8,
        "offset": 0,
      },
      "path": "/blog",
    },
  ],
  [
    {
      "component": "/path/to/blog.js",
      "context": {
        "id": "blog-id",
        "limit": 8,
        "offset": 8,
      },
      "path": "/blog/page/2",
    },
  ],
  [
    {
      "component": "/path/to/blog.js",
      "context": {
        "id": "blog-id",
        "limit": 8,
        "offset": 16,
      },
      "path": "/blog/page/3",
    },
  ],
  [
    {
      "component": "/path/to/blog.js",
      "context": {
        "id": "blog-id",
        "limit": 8,
        "offset": 24,
      },
      "path": "/blog/page/4",
    },
  ],
]
`;

exports[`createPages > 'correctly creates pages (blog-paginated-custom-limit)' > routes.json 1`] = `"[{"name":"blog","path":"/blog","scopes":{"pagination":"/blog/page/:page"}}]"`;

exports[`createPages > 'correctly creates pages (blog-paginated-manual-route)' > createPage 1`] = `
[
  [
    {
      "component": "/path/to/blog.js",
      "context": {
        "id": "blog-id",
        "limit": 10,
        "offset": 0,
      },
      "path": "/blog",
    },
  ],
  [
    {
      "component": "/path/to/blog.js",
      "context": {
        "id": "blog-id",
        "limit": 10,
        "offset": 10,
      },
      "path": "/blog/what/2",
    },
  ],
  [
    {
      "component": "/path/to/blog.js",
      "context": {
        "id": "blog-id",
        "limit": 10,
        "offset": 20,
      },
      "path": "/blog/what/3",
    },
  ],
]
`;

exports[`createPages > 'correctly creates pages (blog-paginated-manual-route)' > routes.json 1`] = `"[{"name":"blog","path":"/blog","scopes":{"pagination":"/blog/what/:page"}},{"name":"blog.paginated","path":"/blog/what/:page","scopes":{}}]"`;

exports[`createPages > 'correctly creates pages (dynamic-page)' > createPage 1`] = `
[
  [
    {
      "component": "/path/to/page.js",
      "context": {
        "id": "page-id",
        "page": "hello",
      },
      "path": "/pages/hello",
    },
  ],
  [
    {
      "component": "/path/to/page.js",
      "context": {
        "id": "page-id",
        "page": "world",
      },
      "path": "/pages/world",
    },
  ],
]
`;

exports[`createPages > 'correctly creates pages (dynamic-page)' > routes.json 1`] = `"[{"name":"page","path":"/pages/:page","scopes":{}}]"`;

exports[`createPages > 'correctly creates pages (dynamic-template-args)' > createPage 1`] = `
[
  [
    {
      "component": "/path/to/page.js?contentPath=content%2Fhello.mdx",
      "context": {
        "id": "page-id",
        "page": "hello",
      },
      "path": "/pages/hello",
    },
  ],
  [
    {
      "component": "/path/to/page.js?contentPath=content%2Fworld.mdx",
      "context": {
        "id": "page-id",
        "page": "world",
      },
      "path": "/pages/world",
    },
  ],
]
`;

exports[`createPages > 'correctly creates pages (dynamic-template-args)' > routes.json 1`] = `"[{"name":"page","path":"/pages/:page","scopes":{}}]"`;

exports[`createPages > 'correctly creates pages (home-about)' > createPage 1`] = `
[
  [
    {
      "component": "/path/to/index.js",
      "context": {
        "id": "home-id",
      },
      "path": "/",
    },
  ],
  [
    {
      "component": "/path/to/page.js",
      "context": {
        "id": "about-id",
      },
      "path": "/about",
    },
  ],
  [
    {
      "component": "/path/to/page.js",
      "context": {
        "id": "about-id",
      },
      "path": "/aboutme",
    },
  ],
]
`;

exports[`createPages > 'correctly creates pages (home-about)' > routes.json 1`] = `"[{"name":"home","path":"/","scopes":{}},{"name":"about","path":"/about","scopes":{}},{"name":"aboutme","path":"/aboutme","scopes":{}}]"`;

exports[`createPages > 'correctly creates pages (page-context)' > createPage 1`] = `
[
  [
    {
      "component": "/path/to/page.js",
      "context": {
        "id": "about-id",
        "name": "adam",
        "profile": {
          "age": 25,
          "gender": "male",
        },
      },
      "path": "/about/adam",
    },
  ],
  [
    {
      "component": "/path/to/page.js",
      "context": {
        "id": "about-id",
        "name": "sara",
        "profile": {
          "age": 28,
          "gender": "female",
        },
      },
      "path": "/about/sara",
    },
  ],
]
`;

exports[`createPages > 'correctly creates pages (page-context)' > routes.json 1`] = `"[{"name":"about","path":"/about/:name","scopes":{}}]"`;

exports[`createPages > 'throws error for (bad-helper-empty)' > error 1`] = `
{
  "error": [Error: Unable to import module "/path/to/helper.js"],
  "text": "The plugin threw an error during "createPages" hook:
 Error while running page helper function at "/path/to/helper.js"",
}
`;

exports[`createPages > 'throws error for (bad-helper-empty-call)' > error 1`] = `
{
  "error": [Error: Route name passed to createAdvancedPage() at "/path/to/helper.js" must be a non-empty string],
  "text": "The plugin threw an error during "createPages" hook",
}
`;

exports[`createPages > 'throws error for (bad-helper-empty-pagination)' > error 1`] = `
{
  "error": [Error: Invalid pagination object passed to createAdvancedPage() at "/path/to/helper.js": "count" parameter is missing],
  "text": "The plugin threw an error during "createPages" hook",
}
`;

exports[`createPages > 'throws error for (bad-helper-invalid-pagination-count)' > error 1`] = `
{
  "error": [Error: Invalid pagination object passed to createAdvancedPage() at "/path/to/helper.js": "count" parameter must be a non-negative integer. Got: "-15"],
  "text": "The plugin threw an error during "createPages" hook",
}
`;

exports[`createPages > 'throws error for (bad-helper-invalid-pagination-limit)' > error 1`] = `
{
  "error": [Error: Invalid pagination object passed to createAdvancedPage() at "/path/to/helper.js": "limit" parameter must be a positive integer. Got: "-10"],
  "text": "The plugin threw an error during "createPages" hook",
}
`;

exports[`createPages > 'throws error for (bad-helper-invalid-pagination-route)' > error 1`] = `
{
  "error": [Error: Invalid pagination object passed to createAdvancedPage() at "/path/to/helper.js": Unrecognized route "foo"],
  "text": "The plugin threw an error during "createPages" hook",
}
`;

exports[`createPages > 'throws error for (bad-helper-invalid-route-params)' > error 1`] = `
{
  "error": [TypeError: Error generating a path for route "/about/:name" with params "{}": Expected "name" to be a string],
  "text": "The plugin threw an error during "createPages" hook:
 Error while running page helper function at "/path/to/helper.js"",
}
`;

exports[`createPages > 'throws error for (bad-helper-invalid-type)' > error 1`] = `
{
  "error": [TypeError: helperFn is not a function],
  "text": "The plugin threw an error during "createPages" hook:
 Error while running page helper function at "/path/to/helper.js"",
}
`;

exports[`createPages > 'throws error for (bad-helper-js-error)' > error 1`] = `
{
  "error": [Error: some error],
  "text": "The plugin threw an error during "createPages" hook:
 Error while running page helper function at "/path/to/helper.js"",
}
`;

exports[`createPages > 'throws error for (bad-helper-undefined-route)' > error 1`] = `
{
  "error": [Error: Unrecognized route "blog" passed to createAdvancedPage() at "/path/to/helper.js"],
  "text": "The plugin threw an error during "createPages" hook",
}
`;

exports[`createPages > 'throws error for (duplicated-routes)' > error 1`] = `
{
  "error": [Error: Unable to create route with name "about". Route name already exists.],
  "text": "The plugin threw an error during "createPages" hook",
}
`;

exports[`createPages > 'throws error for (invalid-helper)' > error 1`] = `
{
  "error": [Error: Unable to import ESM module "/path/to/whatever.js"],
  "text": "The plugin threw an error during "createPages" hook:
 Error while running page helper function at "/path/to/whatever.js"",
}
`;

exports[`createPages > correctly calls graphql pages query 1`] = `
[
  [
    "{
      allPage {
        nodes {
          id
          routes {
            name
            path
          }
          title
          templateName
          template
          helper
        }
      }
    }",
  ],
]
`;

exports[`createPages > correctly handles graphql errors 1`] = `
{
  "error": [Error: Some fatal error],
  "text": "The plugin threw an error during "createPages" hook",
}
`;

exports[`createPages > correctly handles graphql result errors 1`] = `
{
  "error": [Error: Some error],
  "text": "The plugin threw an error during "createPages" hook:
 Failed running "create-pages" GraphQL Query",
}
`;

exports[`createPages > correctly handles routeMap write errors 1`] = `
{
  "error": [Error: ENOENT: no such file or directory,  '/path/to/.cache'],
  "text": "The plugin threw an error during "createPages" hook:
 Error writing route map export file",
}
`;
